File: app/src/main/java/com/example/myapplication/core/model/WorkoutModels.kt
package com.example.myapplication.core.model

data class ActiveGoal(
    val id: Long,
    val config: GoalConfig,
    val totalWorkouts: Int,
)

data class WorkoutSession(
    val id: Long,
    val goalId: Long,
    val sequenceIndex: Int,
    val titleVi: String,
    val focusVi: String,
    val estimatedMinutes: Int,
    val dueEpochDay: Long,
    val exercises: List<WorkoutExercise>,
    val selectedTimeBudgetMinutes: Int? = null,
    val omittedExerciseCount: Int = 0,
)

data class WorkoutExercise(
    val orderIndex: Int,
    val exerciseId: String,
    val prescription: ExercisePrescription,
    val checked: Boolean,
    val originalExerciseId: String? = null,
    val isLightWorkout: Boolean = false,
)

data class CompletedWorkout(
    val goalId: Long,
    val completedEpochDay: Long,
)

data class WorkoutHistoryEntry(
    val sessionId: Long,
    val goalId: Long,
    val sequenceIndex: Int,
    val dueEpochDay: Long,
    val completedEpochDay: Long?,
    val estimatedMinutes: Int,
    val selectedTimeBudgetMinutes: Int? = null,
)

File: flutter/lib/core/model/workout_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'goal_models.dart';
import 'catalog_models.dart';

part 'workout_models.freezed.dart';
part 'workout_models.g.dart';

@freezed
abstract class ActiveGoal with _$ActiveGoal {
  const factory ActiveGoal({
    required int id,
    required GoalConfig config,
    required int totalWorkouts,
  }) = _ActiveGoal;

  factory ActiveGoal.fromJson(Map<String, dynamic> json) =>
      _$ActiveGoalFromJson(json);
}

@freezed
abstract class WorkoutSession with _$WorkoutSession {
  const factory WorkoutSession({
    required int id,
    required int goalId,
    required int sequenceIndex,
    required String titleVi,
    required String focusVi,
    required int estimatedMinutes,
    required int dueEpochDay,
    required List<WorkoutExercise> exercises,
    int? selectedTimeBudgetMinutes,
    @Default(0) int omittedExerciseCount,
  }) = _WorkoutSession;

  factory WorkoutSession.fromJson(Map<String, dynamic> json) =>
      _$WorkoutSessionFromJson(json);
}

@freezed
abstract class WorkoutExercise with _$WorkoutExercise {
  const factory WorkoutExercise({
    required int orderIndex,
    required String exerciseId,
    required ExercisePrescription prescription,
    required bool isChecked,
    String? originalExerciseId,
    @Default(false) bool isLightWorkout,
  }) = _WorkoutExercise;

  factory WorkoutExercise.fromJson(Map<String, dynamic> json) =>
      _$WorkoutExerciseFromJson(json);
}

@freezed
abstract class CompletedWorkout with _$CompletedWorkout {
  const factory CompletedWorkout({
    required int goalId,
    required int completedEpochDay,
  }) = _CompletedWorkout;

  factory CompletedWorkout.fromJson(Map<String, dynamic> json) =>
      _$CompletedWorkoutFromJson(json);
}

@freezed
abstract class WorkoutHistoryEntry with _$WorkoutHistoryEntry {
  const factory WorkoutHistoryEntry({
    required int sessionId,
    required int goalId,
    required int sequenceIndex,
    required int dueEpochDay,
    int? completedEpochDay,
    required int estimatedMinutes,
    int? selectedTimeBudgetMinutes,
  }) = _WorkoutHistoryEntry;

  factory WorkoutHistoryEntry.fromJson(Map<String, dynamic> json) =>
      _$WorkoutHistoryEntryFromJson(json);
}
